### ISSUE TITLE
Waiting for a transaction fails on Geth 1.13.11

### ISSUE THREAD

**Issue by antazoey** [ethereum/web3.py#3212]
* Version: 6.15.0
* Python: 3.10
* OS: macos
* `pip freeze` output

```
<put the output from running pip freeze here>
```

### What was wrong?

The line:

```python
tx = web3.eth.wait_for_transaction_receipt(hex_hash)
```

has issues on my local testing Geth node (inspired from geth --dev).
So I run into this issue if I check too fast:

```
ValueError {"message": "(-32000) transaction indexing is in progress"}
```

### How can it be fixed?

I think raising + catching a custom error for this situation.
Like it should still wait for the timeout if we are indexing.
Right now, it is only looking for "tx not found" error.
---
**Note:** We prefer to use issues to track our work. If you think you've encountered a bug in web3py or 
have a feature request, you're in the right place. If you have implementation or usage questions, 
please refer to our [documentation](https://web3py.readthedocs.io/en/latest/) and/or join the conversation 
on [discord](https://discord.gg/GHryRvPB84).

**Comment by antazoey** [comment]
Realizing this is likely related to a recent Geth change: https://github.com/ethereum/go-ethereum/issues/28877

**Comment by rjl493456442** [comment]
The question is: how does web3.py handle receipt waiting?

My hunch is that there is an internal loop that queries the receipts repeatedly until they are available.

Originally, eth_getTransactionReceipt would return null if the receipt was not available yet. After the implementation of https://github.com/ethereum/go-ethereum/pull/28703, eth_getTransactionReceipt returns an error if the receipt is not available, and transaction indexing is not finished yet (note that transaction indexing only occurs during the initial sync stage).

I suspect that the new behavior breaks the logic in web3.py, which immediately returns when the error is returned by the GETH RPC.

**Comment by fselmo** [comment]
closed by #3216 and #3217
